### ISSUE TITLE
add utilities to create Spans from Futures

### ISSUE THREAD

**Issue by ivantopo** [kamon-io/kamon-futures#3]
There should be a way to facilitate migration from our old [`withNewAsyncSegment`](https://github.com/kamon-io/Kamon/blob/master/kamon-core/src/main/scala/kamon/trace/TraceContext.scala#L65) that can create a Span from a Future.

Thinking about it, it would be great if the Span could start when the future is created and have a mark when it is actually executed but need to dig into the implementation to see if/how if could be possible.

**Comment by jtjeferreira** [comment]
for reference here is the permalink to [withNewAsyncSegment](https://github.com/kamon-io/Kamon/blob/79c7f9bbbbf775e09b7a02109a77986665be43bc/kamon-core/src/main/scala/kamon/trace/TraceContext.scala#L65)

**Comment by jatcwang** [comment]
This is my implementation. Probably not the most efficient due to an additional Future.apply at the start but it works. 

```
  def markSpan[A](operationName: String, extraTags: Map[String, String] = Map.empty)(
    future: => Future[A]
  )(implicit ec: ExecutionContext): Future[A] = {
    Future {
      val builder = Kamon.buildSpan(operationName)
      extraTags.foreach {
        case (key, value) =>
          builder.withTag(key, value)
      }
      val newSpan = builder.start()
      val scope = Kamon.storeContext(Kamon.currentContext().withKey(Span.ContextKey, newSpan))

      try future.transform(
        res => {
          finishSpanAndContextScope(newSpan, scope)
          res
        },
        err => {
          finishSpanAndContextScope(newSpan, scope, err)
          err
        }
      )
      catch {
        case e: Throwable =>
          finishSpanAndContextScope(newSpan, scope, e)
          throw e
      }
    }
  }.flatten

  def finishSpanAndContextScope(span: Span, scope: Storage.Scope) = {
    span.finish()
    scope.close()
  }

  def finishSpanAndContextScope(span: Span, scope: Storage.Scope, throwable: Throwable) = {
    span.addError(throwable.getMessage, throwable)
    span.finish()
    scope.close()
  }
```

Or we can implement our own alternative `Future.apply` (not tested):

```
def markFuture[A](body: => A): Future[A] = {
  Future.unit.map { _ =>
    Kamon.withSpan(...) {
      body
    }  
  }
}
```

**Comment by jalaziz** [comment]
Delayed span seems to get us halfway there. However, it would still be great to have something that is equivalent to `withNewAsyncSegment`.

I should add, `traceAsync` is confusingly named. The name would suggest it can be used to trace async functions, but it doesn't seem like it can. The only difference from `trace` appears to be that it accepts a parameter.

**Comment by ivantopo** [comment]
> I should add, traceAsync is confusingly named. 

I totally agree with you. If it was possible I would have both functions named `trace` but their return types are different :confused:.. it was previously called `tracedCallback` which seemed weird.. any suggestions on a better name would be appreciated! 

Regarding `withNewAsyncSegment`, it seems like a combination of @jatcwang's solution above and @jtjeferreira's https://github.com/kamon-io/kamon-futures/pull/5/files should be the way to go. I just noticed that those solutions and the one that is currently included in instrumentation are aiming for two different use cases:
  - The solutions discussed here are a much simpler and general approach to instrumenting one Future, which could have a bunch of chained transformations but the attention is focused on this one "wrapper" future that represents the whole computation and how much time did it take to happen.
  - The solution included in the module now is not focused on the Future itself but rather on each of the steps that make up the final Future (the body and then any map/flatmap/xxx) but does not allow you to get the time of the whole computation. 

I totally missed the fact that both are needed here (thanks a lot for bringing up this discussion!). Now, it seems like we need to end up with 3 different functions:
  - one that tracks a Future, regardless of its internal transformations (if any)
  - one that tracks a Future's body
  - one that tracks a Future's transformations (which is essentially the same as above, but types are slightly different)

Now, the question of the day: what should be the names of these?.. I'm leaning to think that the whole Future option is going to be more commonly used than the individual transformations variants so I would propose:
  - `trace[T](name.....)(f: => Future[T]): Future[T]`: to create a Span with the provided info and finish that Span when the Future is finished. This would be a regular Span.
  - `traceBody[T](name...)(body: => T): T`: to create a Span out of a Future's body. This would be a delayed Span.
  - `traceFun[T, S](name...)(fun: T => S): T => S`: to create a Span out of a Future's transformation. This would be a delayed Span.

What do you guys think about these options?

**Comment by jalaziz** [comment]
I believe you can still name all three `trace` since the curried parameters are different.

**Comment by ivantopo** [comment]
Folks, I just put together the changes mentioned above on https://github.com/kamon-io/kamon-futures/pull/12, feedback will be appreciated :)

@jatcwang taking a closer look at your example, I noticed that in your code the Scope created when setting the Span as current is being closed from a different Thread (it happens inside the transform callbacks) and that is something that can definitely cause visibility issues (the original Thread might not see the changes to the Context storage since the change was triggered from another Thread). Please take into account that all Scopes created in one Thread should be closed from within that same Thread!

**Comment by jatcwang** [comment]
Hi @ivantopo. Thanks for keeping me in the loop. I'll take a look over the next few days :)
